arm9/source/gamecart/gamecart.c

```diff
@@ -221,3 +239,31 @@ u32 ReadCartPrivateHeader(void* buffer, u64 offset, u64 count, CartData* cdata)
     }
     return 0;
 }
+
+u32 ReadCartSave(u8* buffer, u64 offset, u64 count, CartData* cdata) {
+    if (offset >= cdata->save_size) return 1;
+    if (offset + count > cdata->save_size) count = cdata->save_size - offset;
+    return (CardSPIReadSaveData(cdata->save_type, offset, buffer, count) == 0) ? 0 : 1;
+}
+
+u32 WriteCartSave(const u8* buffer, u64 offset, u64 count, CartData* cdata) {
+    if (offset >= cdata->save_size) return 1;
+    if (offset + count > cdata->save_size) count = cdata->save_size - offset;
+    return (CardSPIWriteSaveData(cdata->save_type, offset, buffer, count) == 0) ? 0 : 1;
+}
+
+u32 ReadCartSaveJedecId(u8* buffer, u64 offset, u64 count, CartData* cdata) {
+    u8 ownBuf[JEDECID_AND_SREG_SIZE] = { 0 };
+    u32 id;
+    u8 sReg;
+    if (offset >= JEDECID_AND_SREG_SIZE) return 1;
+    if (offset + count > JEDECID_AND_SREG_SIZE) count = JEDECID_AND_SREG_SIZE - offset;
+    int res = CardSPIReadJEDECIDAndStatusReg(cdata->save_type, &id, &sReg);
+    if (res) return res;
+    ownBuf[0] = (id >> 16) & 0xff;
+    ownBuf[1] = (id >> 8) & 0xff;
+    ownBuf[2] = id & 0xff;
+    ownBuf[JEDECID_AND_SREG_SIZE - 1] = sReg;
+    memcpy(buffer, ownBuf, count);
```

Comment:
Shouldn't this be `memcpy(buffer, ownBuf + offset, count);`? I know it'll be very rare to do offset reads on a (currently) 4 byte virtual file, but regardless this looks wrong.

Reply:
Didn't catch that, thank you!